File: clients/python/provider/superposition_provider/interfaces.py
"""
Interface definitions for feature resolution and experiment metadata.

These interfaces define the core capabilities for flag evaluation and variant resolution.
"""

import logging
from abc import ABC, abstractmethod
from typing import Dict, List, Optional, Any, Union, Mapping, Sequence

from openfeature.evaluation_context import EvaluationContext
from openfeature.exception import ErrorCode
from openfeature.flag_evaluation import FlagResolutionDetails

logger = logging.getLogger(__name__)


class FeatureExperimentMeta(ABC):
    """Trait for experiment variant resolution.

    Implementors provide the ability to determine which experiment variants
    are applicable for a given evaluation context.
    """

    @abstractmethod
    async def get_applicable_variants(
        self,
        context: Optional[EvaluationContext],
        prefix_filter: Optional[List[str]] = None,
    ) -> List[str]:
        """Get the list of applicable experiment variant IDs for the given context.

        Args:
            context: Evaluation context with targeting key and attributes.
            prefix_filter: Optional list of variant ID prefixes to filter by.

        Returns:
            List of variant IDs that apply to this context.
        """
        pass


class AllFeatureProvider(ABC):
    """Trait for bulk configuration resolution.

    Implementors provide the ability to resolve all feature flags at once,
    optionally filtered by key prefixes.
    """

    def resolve_all_features(
        self,
        context: Optional[EvaluationContext],
    ) -> Dict[str, Any]:
        """Resolve all features for the given evaluation context.

        Args:
            context: Evaluation context with targeting key and attributes.

        Returns:
            Map of all resolved feature flags and their values.
        """
        return self.resolve_all_features_with_filter(context, None)

    @abstractmethod
    def resolve_all_features_with_filter(
        self,
        context: Optional[EvaluationContext],
        prefix_filter: Optional[List[str]] = None,
    ) -> Dict[str, Any]:
        """Resolve all features, optionally filtered by key prefixes.

        Args:
            context: Evaluation context with targeting key and attributes.
            prefix_filter: Optional list of key prefixes to include.

        Returns:
            Map of filtered resolved feature flags and their values.
        """
        pass

    def resolve_typed(
        self,
        flag_key: str,
        evaluation_context: Optional[EvaluationContext],
        type_name: str,
        extractor,
        default
    ) -> FlagResolutionDetails:
        """Generic typed resolution using an extractor function.

        Resolves all features and extracts a specific flag, applying type conversion.

        Args:
            flag_key: The flag key to resolve.
            evaluation_context: Evaluation context.
            type_name: Type name for error messages.
            extractor: Function to extract the correct type from the value.
            default: The default value to use if the flag key is not found.

        Returns:
            FlagResolutionDetails with the typed value or error.
        """
        try:
            config = self.resolve_all_features(evaluation_context)
            if flag_key not in config:
                return FlagResolutionDetails(default)

            value = config[flag_key]
            extracted = extractor(value)
            if extracted is None:
                return FlagResolutionDetails(default)
            return FlagResolutionDetails(extracted)
        except Exception as e:
            logger.error(f"Error evaluating {type_name} flag {flag_key}: {e}")
            return FlagResolutionDetails(
                default,
                error_code=ErrorCode.GENERAL,
                error_message=f"Error evaluating flag '{flag_key}': {e}"
            )

    def resolve_bool(
        self,
        flag_key: str,
        default_value: bool,
        evaluation_context: Optional[EvaluationContext],
    ) -> FlagResolutionDetails[bool]:
        """Resolve a boolean flag."""
        return self.resolve_typed(
            flag_key,
            evaluation_context,
            "boolean",
            lambda v: _to_bool(v),
            default_value
        )

    def resolve_string(
        self,
        flag_key: str,
        default_value: str,
        evaluation_context: Optional[EvaluationContext],
    ) -> FlagResolutionDetails[str]:
        """Resolve a string flag."""
        return self.resolve_typed(
            flag_key,
            evaluation_context,
            "string",
            lambda v: v if isinstance(v, str) else None,
            default_value
        )

    def resolve_int(
        self,
        flag_key: str,
        default_value: int,
        evaluation_context: Optional[EvaluationContext],
    ) -> FlagResolutionDetails[int]:
        """Resolve an integer flag."""
        return self.resolve_typed(
            flag_key,
            evaluation_context,
            "integer",
            lambda v: v if isinstance(v, int) and not isinstance(v, bool) else None,
            default_value
        )

    def resolve_float(
        self,
        flag_key: str,
        default_value: float,
        evaluation_context: Optional[EvaluationContext],
    ) -> FlagResolutionDetails[float]:
        """Resolve a float flag."""
        return self.resolve_typed(
            flag_key,
            evaluation_context,
            "float",
            lambda v: v if isinstance(v, (int, float)) and not isinstance(v, bool) else None,
            default_value
        )

    def resolve_object(
        self,
        flag_key: str,
        default_value: Union[Mapping, Sequence],
        evaluation_context: Optional[EvaluationContext],
    ) -> FlagResolutionDetails[Union[Mapping, Sequence]]:
        """Resolve a struct/object flag."""
        return self.resolve_typed(
            flag_key,
            evaluation_context,
            "struct",
            lambda v: v if isinstance(v, dict) or isinstance(v, list) else None,
            default_value
        )

    async def resolve_all_features_async(
        self,
        context: Optional[EvaluationContext],
    ) -> Dict[str, Any]:
        """Resolve all features for the given evaluation context.

        Args:
            context: Evaluation context with targeting key and attributes.

        Returns:
            Map of all resolved feature flags and their values.
        """
        return await self.resolve_all_features_with_filter_async(context, None)

    @abstractmethod
    async def resolve_all_features_with_filter_async(
        self,
        context: Optional[EvaluationContext],
        prefix_filter: Optional[List[str]] = None,
    ) -> Dict[str, Any]:
        """Resolve all features, optionally filtered by key prefixes.

        Args:
            context: Evaluation context with targeting key and attributes.
            prefix_filter: Optional list of key prefixes to include.

        Returns:
            Map of filtered resolved feature flags and their values.
        """
        pass

    async def resolve_typed_async(
        self,
        flag_key: str,
        evaluation_context: Optional[EvaluationContext],
        type_name: str,
        extractor,
        default
    ) -> FlagResolutionDetails:
        """Generic typed resolution using an extractor function.

        Resolves all features and extracts a specific flag, applying type conversion.

        Args:
            flag_key: The flag key to resolve.
            evaluation_context: Evaluation context.
            type_name: Type name for error messages.
            extractor: Function to extract the correct type from the value.
            default: The default value to return if the type cannot be extracted.

        Returns:
            FlagResolutionDetails with the typed value or error.
        """
        try:
            config = await self.resolve_all_features_async(evaluation_context)
            if flag_key not in config:
                return FlagResolutionDetails(default)

            value = config[flag_key]
            extracted = extractor(value)
            if extracted is None:
                return FlagResolutionDetails(default)
            return FlagResolutionDetails(extracted)
        except Exception as e:
            logger.error(f"Error evaluating {type_name} flag {flag_key}: {e}")
            return FlagResolutionDetails(
                default,
                error_code=ErrorCode.GENERAL,
                error_message=f"Error evaluating flag '{flag_key}': {e}"
            )

    async def resolve_bool_async(
        self,
        flag_key: str,
        default_value: bool,
        evaluation_context: Optional[EvaluationContext],
    ) -> FlagResolutionDetails[bool]:
        """Resolve a boolean flag."""
        return await self.resolve_typed_async(
            flag_key,
            evaluation_context,
            "boolean",
            lambda v: _to_bool(v),
            default_value
        )

    async def resolve_string_async(
        self,
        flag_key: str,
        default_value: str,
        evaluation_context: Optional[EvaluationContext],
    ) -> FlagResolutionDetails[str]:
        """Resolve a string flag."""
        return await self.resolve_typed_async(
            flag_key,
            evaluation_context,
            "string",
            lambda v: v if isinstance(v, str) else None,
            default_value
        )

    async def resolve_int_async(
        self,
        flag_key: str,
        default_value: int,
        evaluation_context: Optional[EvaluationContext],
    ) -> FlagResolutionDetails[int]:
        """Resolve an integer flag."""
        return await self.resolve_typed_async(
            flag_key,
            evaluation_context,
            "integer",
            lambda v: v if isinstance(v, int) and not isinstance(v, bool) else None,
            default_value
        )

    async def resolve_float_async(
        self,
        flag_key: str,
        default_value: float,
        evaluation_context: Optional[EvaluationContext],
    ) -> FlagResolutionDetails[float]:
        """Resolve a float flag."""
        return await self.resolve_typed_async(
            flag_key,
            evaluation_context,
            "float",
            lambda v: v if isinstance(v, (int, float)) and not isinstance(v, bool) else None,
            default_value
        )

    async def resolve_object_async(
        self,
        flag_key: str,
        default_value: Union[Mapping, Sequence],
        evaluation_context: Optional[EvaluationContext],
    ) -> FlagResolutionDetails[Union[Mapping, Sequence]]:
        """Resolve a struct/object flag."""
        return await self.resolve_typed_async(
            flag_key,
            evaluation_context,
            "struct",
            lambda v: v if isinstance(v, dict) or isinstance(v, list) else None,
            default_value
        )

def _to_bool(value: Any) -> Optional[bool]:
    if isinstance(value, bool): return value
    if isinstance(value, str):
        if value.lower() == "true": return True
        if value.lower() == "false": return False
    if isinstance(value, (int, float)): return value != 0
    return None

File: clients/javascript/open-feature-provider/interfaces.ts
/**
 * Shared resolution contract for the Superposition providers.
 *
 * `AllFeatureProvider` resolves all features once, extracts a single flag, and applies the unified
 * type-coercion contract (FLAG_NOT_FOUND / TYPE_MISMATCH / GENERAL). Experiment-variant resolution is
 * a separate concern, kept in its own {@link FeatureExperimentMeta} interface to match the reference
 * clients' trait split.
 * OpenFeature JS resolvers may be async, so there is a single async path.
 *
 * The FFI hands back real typed values (numbers, booleans, objects, arrays), so the extractors test
 * the JS type directly rather than decoding per key.
 */

import {
    EvaluationContext,
    ErrorCode,
    JsonValue,
    Logger,
    OpenFeatureEventEmitter,
    ProviderMetadata,
    ResolutionDetails,
    StandardResolutionReasons,
    Hook,
} from "@openfeature/server-sdk";

/** Extracts a typed value from a resolved flag value, or `undefined` on a type mismatch. */
type Extractor<T> = (value: any) => T | undefined;

const asBoolean: Extractor<boolean> = (v) =>
    typeof v === "boolean" ? v : undefined;
const asString: Extractor<string> = (v) =>
    typeof v === "string" ? v : undefined;
// `typeof true === "boolean"`, so booleans never leak in as numbers.
const asNumber: Extractor<number> = (v) =>
    typeof v === "number" && !Number.isNaN(v) ? v : undefined;
// Arrays and non-null objects both count as "object" (arrays allowed, matching Python/JS).
const asObject: Extractor<any> = (v) =>
    typeof v === "object" && v !== null ? v : undefined;

/**
 * Experiment-variant resolution. This is a capability distinct from feature resolution — the two are
 * separate traits in the Rust client (`FeatureExperimentMeta` vs `AllFeatureProvider`) and separate
 * ABCs in Python — so it lives in its own interface here rather than being folded into the base class.
 * Concrete providers implement it alongside extending {@link AllFeatureProvider}.
 */
export interface FeatureExperimentMeta {
    /** Get the applicable experiment variant IDs for the given context. */
    getApplicableVariants(
        context: EvaluationContext,
        prefixFilter?: string[],
        excludePrefixFilter?: string[],
    ): Promise<string[]>;
}

/**
 * The bulk-resolution contract (the Rust trait / Python ABC / Java interface of the same name):
 * resolve all features once, extract a single flag, and apply the unified type-coercion contract. It
 * supplies the shared OpenFeature `resolve*Evaluation` implementations and provider plumbing, but does
 * not itself declare `implements Provider` — it is not a usable provider on its own (no
 * `initialize`/`status`). The concrete providers extend this, declare `implements Provider`, and
 * separately implement {@link FeatureExperimentMeta} — mirroring the reference clients, where
 * `AllFeatureProvider` is a distinct trait/mixin from the OpenFeature provider it is combined into.
 */
export abstract class AllFeatureProvider {
    abstract readonly metadata: ProviderMetadata;

    readonly runsOn = "server" as const;
    events = new OpenFeatureEventEmitter();
    readonly hooks: Hook[] = [];

    /**
     * Resolve all features for the given context, optionally filtered by key prefixes. Concrete
     * providers implement this (locally via the FFI cache, or remotely via the API).
     */
    abstract resolveAllFeaturesWithFilter(
        context: EvaluationContext,
        prefixFilter?: string[],
        excludePrefixFilter?: string[],
    ): Promise<Record<string, any>>;

    /** Resolve all features for the given context (no prefix filter). */
    resolveAllFeatures(
        context: EvaluationContext,
    ): Promise<Record<string, any>> {
        return this.resolveAllFeaturesWithFilter(context, undefined, undefined);
    }

    /**
     * Resolve all features, extract `flagKey`, and apply `extractor`.
     *
     * `reason` is left unset on success — reporting it accurately (STATIC / TARGETING_MATCH / SPLIT)
     * needs the core to say, per key, where a value came from; until it does, guessing would be worse
     * than saying nothing. The same TODO applies to the Rust/Python/Java clients.
     */
    protected async resolveTyped<T>(
        flagKey: string,
        context: EvaluationContext,
        typeName: string,
        extractor: Extractor<T>,
        defaultValue: T,
        logger?: Logger,
    ): Promise<ResolutionDetails<T>> {
        try {
            const config = await this.resolveAllFeatures(context);
            if (!(flagKey in config)) {
                return {
                    value: defaultValue,
                    reason: StandardResolutionReasons.ERROR,
                    errorCode: ErrorCode.FLAG_NOT_FOUND,
                    errorMessage: `Flag '${flagKey}' not found`,
                };
            }
            const extracted = extractor(config[flagKey]);
            if (extracted === undefined) {
                return {
                    value: defaultValue,
                    reason: StandardResolutionReasons.ERROR,
                    errorCode: ErrorCode.TYPE_MISMATCH,
                    errorMessage: `Flag '${flagKey}' is not a ${typeName}`,
                };
            }
            return { value: extracted };
        } catch (error) {
            // Log the evaluation failure through OpenFeature's logger (the SDK passes one per call),
            // mirroring the reference clients which log it in resolve_typed before returning GENERAL.
            const message = `Error evaluating ${typeName} flag '${flagKey}': ${
                error instanceof Error ? error.message : String(error)
            }`;
            logger?.error(message);
            return {
                value: defaultValue,
                reason: StandardResolutionReasons.ERROR,
                errorCode: ErrorCode.GENERAL,
                errorMessage: message,
            };
        }
    }

    // --- OpenFeature FeatureProvider methods ---

    resolveBooleanEvaluation(
        flagKey: string,
        defaultValue: boolean,
        context: EvaluationContext,
        logger?: Logger,
    ): Promise<ResolutionDetails<boolean>> {
        return this.resolveTyped(
            flagKey,
            context,
            "boolean",
            asBoolean,
            defaultValue,
            logger,
        );
    }

    resolveStringEvaluation(
        flagKey: string,
        defaultValue: string,
        context: EvaluationContext,
        logger?: Logger,
    ): Promise<ResolutionDetails<string>> {
        return this.resolveTyped(
            flagKey,
            context,
            "string",
            asString,
            defaultValue,
            logger,
        );
    }

    resolveNumberEvaluation(
        flagKey: string,
        defaultValue: number,
        context: EvaluationContext,
        logger?: Logger,
    ): Promise<ResolutionDetails<number>> {
        return this.resolveTyped(
            flagKey,
            context,
            "number",
            asNumber,
            defaultValue,
            logger,
        );
    }

    resolveObjectEvaluation<T extends JsonValue>(
        flagKey: string,
        defaultValue: T,
        context: EvaluationContext,
        logger?: Logger,
    ): Promise<ResolutionDetails<T>> {
        return this.resolveTyped<T>(
            flagKey,
            context,
            "object",
            asObject,
            defaultValue,
            logger,
        );
    }
}
